File: apps/android/app/src/main/java/pro/wewed/app/services/ServerBackedWeddingDayOperations.kt
package pro.wewed.app.services

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import pro.wewed.app.models.AnnouncementUrgency
import pro.wewed.app.models.VendorPresence
import pro.wewed.app.models.VendorPresenceState
import pro.wewed.app.models.WeddingAnnouncement
import java.util.UUID

/**
 * Server-backed Wedding Day API client for Android.
 * Connects to real Next.js API routes in isolatedWeddingDay or production modes.
 * All operations degrade gracefully on network failure — callers should fall back to [FixtureWeddingRepository].
 */
class ServerBackedWeddingDayOperations(
    private val transport: WeddingDayHttpTransport,
    private val bearerToken: String,
    private val weddingId: String
) {
    private val gson = Gson()
    private val authHeader get() = mapOf(
        "Authorization" to "Bearer $bearerToken",
        "Accept" to "application/json"
    )

    // ── Announcements ────────────────────────────────────────────────────────

    suspend fun getAnnouncements(): List<WeddingAnnouncement> {
        return try {
            val resp = transport.get("api/wedding-day/announcements", authHeader)
            if (resp.status !in 200..299) return emptyList()
            val type = object : TypeToken<AnnouncementEnvelope>() {}.type
            val envelope: AnnouncementEnvelope = gson.fromJson(resp.body, type)
            if (!envelope.success) return emptyList()
            envelope.data.map { item ->
                WeddingAnnouncement(
                    id = item.id ?: UUID.randomUUID().toString(),
                    title = item.title ?: "Announcement",
                    message = item.body,
                    urgency = if (item.urgency == "high") AnnouncementUrgency.ALERT else AnnouncementUrgency.INFO
                )
            }
        } catch (_: Exception) {
            emptyList()
        }
    }

    suspend fun postAnnouncement(
        title: String,
        message: String,
        urgency: AnnouncementUrgency
    ): WeddingAnnouncement? {
        return try {
            val headers = authHeader + mapOf("Content-Type" to "application/json")
            val body = gson.toJson(mapOf("title" to title, "body" to message, "audience" to "all"))
            val resp = transport.post("api/wedding-day/announcements", headers, body)
            if (resp.status !in 200..299) return null
            val type = object : TypeToken<CreateAnnouncementEnvelope>() {}.type
            val envelope: CreateAnnouncementEnvelope = gson.fromJson(resp.body, type)
            if (!envelope.success) return null
            WeddingAnnouncement(
                id = envelope.data.id,
                title = envelope.data.title ?: title,
                message = envelope.data.body,
                urgency = urgency
            )
        } catch (_: Exception) {
            null
        }
    }

    // ── Vendor Presence ──────────────────────────────────────────────────────

    suspend fun updateVendorState(
        serviceEngagementId: String,
        state: VendorPresenceState
    ): VendorPresence? {
        return try {
            val stateString = when (state) {
                VendorPresenceState.SCHEDULED -> "CONFIRMED"
                VendorPresenceState.EN_ROUTE -> "EN_ROUTE"
                VendorPresenceState.ARRIVED -> "ARRIVED_ON_SITE"
                VendorPresenceState.SERVICE_ACTIVE -> "SERVICE_ACTIVE"
                VendorPresenceState.COMPLETED -> "COMPLETED"
            }
            val headers = authHeader + mapOf("Content-Type" to "application/json")
            val body = gson.toJson(mapOf("serviceEngagementId" to serviceEngagementId, "state" to stateString))
            val resp = transport.post("api/wedding-day/vendors/presence", headers, body)
            if (resp.status !in 200..299) return null
            val type = object : TypeToken<PresenceEnvelope>() {}.type
            val envelope: PresenceEnvelope = gson.fromJson(resp.body, type)
            if (!envelope.success) return null
            VendorPresence(
                id = serviceEngagementId,
                vendorName = "Vendor Service",
                serviceCategory = "Service Provider",
                serviceArea = "Main Venue",
                state = state,
                expectedTime = "On Site"
            )
        } catch (_: Exception) {
            null
        }
    }

    // ── Internal DTOs ────────────────────────────────────────────────────────

    private data class AnnouncementEnvelope(val success: Boolean, val data: List<AnnouncementPayload>)
    private data class AnnouncementPayload(val id: String?, val title: String?, val body: String, val urgency: String?)
    private data class CreateAnnouncementEnvelope(val success: Boolean, val data: CreatedItem)
    private data class CreatedItem(val id: String, val title: String?, val body: String)
    private data class PresenceEnvelope(val success: Boolean)
}

File: apps/ios/Wewed/Services/ServerBackedWeddingDayOperations.swift
import Foundation

/// Server-backed operations for Wedding Day APIs on iOS.
/// Connects to real Next.js API routes when operating in isolatedWeddingDay or production modes.
public actor ServerBackedWeddingDayOperations: Sendable {
    private let baseURL: URL
    private let bearerToken: String
    private let weddingId: String
    private let session: URLSession
    private let decoder = JSONDecoder()

    public init(
        baseURL: URL,
        bearerToken: String,
        weddingId: String,
        session: URLSession = .shared
    ) {
        self.baseURL = baseURL
        self.bearerToken = bearerToken
        self.weddingId = weddingId
        self.session = session
    }

    /// Fetches the guest's Wedding Pass from the server-backed endpoint.
    public func getWeddingPass(token: String) async throws -> WeddingPass? {
        let url = baseURL.appendingPathComponent("api/wedding-day/pass")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        if !bearerToken.isEmpty {
            request.setValue("Bearer \(bearerToken)", forHTTPHeaderField: "Authorization")
        }
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            return nil
        }

        struct PassEnvelope: Codable {
            let success: Bool
            let data: PassPayload
        }
        struct PassPayload: Codable {
            let id: String
            let weddingId: String
            let guestId: String
            let passSerial: String
            let token: String
        }

        guard let envelope = try? decoder.decode(PassEnvelope.self, from: data), envelope.success else {
            return nil
        }

        return WeddingPass(
            token: envelope.data.token,
            weddingId: envelope.data.weddingId,
            coupleNames: "Wedding Couple",
            weddingDate: ISO8601DateFormatter().string(from: Date()),
            venueName: "Ceremony Venue",
            venueAddress: "Harare, Zimbabwe",
            guestName: "Valued Guest",
            householdName: nil,
            partySize: 1,
            tableNumber: nil,
            tableName: nil,
            seatNumber: nil,
            currentStage: .attending,
            qrPayload: envelope.data.token
        )
    }

    /// Fetches active announcements from the server.
    public func getAnnouncements() async throws -> [WeddingAnnouncement] {
        let url = baseURL.appendingPathComponent("api/wedding-day/announcements")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        if !bearerToken.isEmpty {
            request.setValue("Bearer \(bearerToken)", forHTTPHeaderField: "Authorization")
        }
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            return []
        }

        struct AnnouncementEnvelope: Codable {
            let success: Bool
            let data: [AnnouncementPayload]
        }
        struct AnnouncementPayload: Codable {
            let id: String?
            let title: String?
            let body: String
            let urgency: String?
            let createdAt: String?
        }

        guard let envelope = try? decoder.decode(AnnouncementEnvelope.self, from: data), envelope.success else {
            return []
        }

        return envelope.data.map { item in
            WeddingAnnouncement(
                id: item.id ?? UUID().uuidString,
                title: item.title ?? "Announcement",
                message: item.body,
                urgency: item.urgency == "high" ? .alert : .info,
                timestamp: Date()
            )
        }
    }

    /// Publishes a wedding day announcement.
    public func postAnnouncement(
        title: String,
        message: String,
        urgency: AnnouncementUrgency
    ) async throws -> WeddingAnnouncement? {
        let url = baseURL.appendingPathComponent("api/wedding-day/announcements")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(bearerToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let payload: [String: Any] = [
            "title": title,
            "body": message,
            "audience": "all"
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            return nil
        }

        struct CreateEnvelope: Codable {
            let success: Bool
            let data: CreatedItem
        }
        struct CreatedItem: Codable {
            let id: String
            let title: String?
            let body: String
        }

        guard let envelope = try? decoder.decode(CreateEnvelope.self, from: data), envelope.success else {
            return nil
        }

        return WeddingAnnouncement(
            id: envelope.data.id,
            title: envelope.data.title ?? title,
            message: envelope.data.body,
            urgency: urgency,
            timestamp: Date()
        )
    }

    /// Transitions vendor presence on site.
    public func updateVendorState(
        serviceEngagementId: String,
        state: VendorPresenceState
    ) async throws -> VendorPresence? {
        let url = baseURL.appendingPathComponent("api/wedding-day/vendors/presence")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(bearerToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let stateString: String
        switch state {
        case .scheduled: stateString = "CONFIRMED"
        case .enRoute: stateString = "EN_ROUTE"
        case .arrived: stateString = "ARRIVED_ON_SITE"
        case .serviceActive: stateString = "SERVICE_ACTIVE"
        case .completed: stateString = "COMPLETED"
        }

        let payload: [String: Any] = [
            "serviceEngagementId": serviceEngagementId,
            "state": stateString
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            return nil
        }

        struct PresenceEnvelope: Codable {
            let success: Bool
        }
        guard let envelope = try? decoder.decode(PresenceEnvelope.self, from: data), envelope.success else {
            return nil
        }

        return VendorPresence(
            id: serviceEngagementId,
            vendorName: "Vendor Service",
            serviceCategory: "Service Provider",
            serviceArea: "Main Venue",
            state: state,
            expectedTime: "On Site",
            lastUpdated: Date()
        )
    }
}
